modules/packages/Channel.chpl

```diff
@@ -0,0 +1,574 @@
+    * :proc:`recv` : Receive a value from the channel.
+    * :proc:`close` : Close a channel such that no more value can be sent to it.
+
+The channel operations are blocking, i.e., the calling task will be suspended
+if an operation cannot be completed. The channel follows First-In-First-Out
+mechanism, i.e., the first value sent to the channel will be received first.
+*/
+
+module Channel {
+
+    use CPtr;
+    use Sort;
+    use SysCTypes;
+    use List;
+    use Random;
+
+    pragma "no doc"
+    class Waiter {
+        // Class used to maintain the properties of suspended tasks.
+
+        type valueType;
+        var val : c_ptr(valueType);
+        var processPtr : c_ptr(single bool);
+        var isSelect : bool;
+
+        // These variables are specific to the waiters for select statement
+        var isSelectDone : c_ptr(atomic int);
+        var selectId : int = -1;
+
+        var prev : unmanaged Waiter(valueType)?;
+        var next : unmanaged Waiter(valueType)?;
+
+        proc init(ref value, ref process$ : single bool) {
+            valueType = value.type;
+            val = c_ptrTo(value);
+            processPtr = c_ptrTo(process$);
+            isSelect = false;
+        }
+
+        proc init(ref value, ref process$ : single bool, ref selectDone : atomic int, caseId : int) {
+            valueType = value.eltType;
+            val = value;
+            processPtr = c_ptrTo(process$);
+            isSelect = true;
+            isSelectDone = c_ptrTo(selectDone);
+            selectId = caseId;
+        }
+
+        proc suspend() : bool {
+            return processPtr.deref().readFF();
+        }
+
+        proc release(status : bool) {
+            processPtr.deref().writeEF(status);
+        }
+    }
+
+    /*
+    Implementation of doubly-ended queue to keep track of suspended receiving
+    and sending tasks.
+    */
+    pragma "no doc"
+    class WaiterQue {
+        type eltType;
+        var front : unmanaged Waiter(eltType)?;
+        var back : unmanaged Waiter(eltType)?;
+
+        /* Push the `waiter` into the queue */
+        proc enque(waiter : unmanaged Waiter(eltType)) {
+            if front == nil {
+                front = waiter;
+                back = waiter;
+            }
+            else {
+                back!.next = waiter;
+                waiter.prev = back;
+                back = waiter;
+            }
+        }
+
+        proc isEmpty() : bool{
+            return (front == nil);
+        }
+
+        /* Pop the first value from the queue */
+        proc deque() : unmanaged Waiter(eltType) {
+            var waiter : unmanaged Waiter(eltType)?;
+            if front == back {
+                waiter = front;
+                front = nil;
+                back = nil;
+            }
+            else {
+                waiter = front;
+                front = front!.next;
+                front!.prev = nil;
+                waiter!.next = nil;
```

Comment:
This sets waiter.next to nil, but waiter.prev might be left with a value. Additionally, if clearing these prev/next pointers is important, should it be also done in the other cases of the following function to dequeue a specific entry?

Reply:
In this case, the waiter.prev would already be `nil` because the waiter is at the front of the queue. Clearing the pointers is important in this function because when calling the `dequeue` with parameters, the specified waiter may be already dequeued from the queue. In that case, the function would not do anything.

Reply:
Right, in that case, shouldn't the `dequeue` with parameters also clear these fields?

Reply:
I.e. `proc dequeue(waiter)` could end with `waiter.prev = nil; waiter.next = nil;`